### PR TITLE
Supporting logits as parameters in Bernoulli and Categorical

### PR DESCRIPTION
**Pull request by neerajprad**
This allows for using logits directly as parameters in Bernoulli and Categorical distributions. 

**Details:**
 - Uses the binary cross entropy operation to score Bernoulli samples. This makes the `log_prob` computation differentiable.
 - Categorical normalizes the probs/logits parameters provided in the constructor and stores these as attributes, so that other methods do not need to bother with normalizing unnormalized probabilities/logits.
 - Simplifies entropy computation (and corrects it for the case of Categorical with unnormalized probability values as parameters).

### CODE REVIEW

**Comment by neerajprad** [comment]
cc. @apaszke

File: torch/distributions/bernoulli.py [hunk 1]
```diff
@@ -37,20 +45,17 @@ def sample(self, sample_shape=torch.Size()):
...
+
     def log_prob(self, value):
         self._validate_log_prob_arg(value)
-        param_shape = value.size()
-        probs = self.probs.expand(param_shape)
-        # compute the log probabilities for 0 and 1
-        log_pmf = (torch.stack([1 - probs, probs], dim=-1)).log()
-        # evaluate using the values
-        return log_pmf.gather(-1, value.unsqueeze(-1).long()).squeeze(-1)
+        return -self._binary_cross_entropy(value)
```
Comment (fritzo) [inline torch/distributions/bernoulli.py]: Is it still numerically stable if we use the [non-reducing C implementation](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1235) with `probs` instead of `logits`?
```py
return -torch.nn.functiona.binary_cross_entropy(self.probs, value, reduce=False)
```
(Note that `torch.nn.functional.binary_cross_entropy_with_logits()` does not allow `reduce=False`)

Comment (neerajprad) [reply torch/distributions/bernoulli.py]: I wanted to use `cross_entropy_with_logits` but you are right that `reduce` arg is not available with that. 

If we are operating with logits, this will be more stable. See:

```
>>> b = Bernoulli(logits=torch.FloatTensor([-32, -8]))
>>> b.log_prob(torch.FloatTensor([0, 1]))

-0.0000
-8.0003
[torch.FloatTensor of size 2]

>>> torch._C._nn.binary_cross_entropy(Variable(b.probs), Variable(torch.FloatTensor([0, 1])), reduce=False)
Variable containing:
-9.8743e-13
 8.0003e+00
[torch.FloatTensor of size 2]
```

Comment (fritzo) [reply torch/distributions/bernoulli.py]: Great, just checking. It might be cleaner to add a `reduce=True` argument to `torch.nn.functional.binary_cross_entropy_with_logits()` now that we're moving this code into PyTorch.

Comment (neerajprad) [reply torch/distributions/bernoulli.py]: Good idea, will do that instead.

File: torch/distributions/bernoulli.py [hunk 2]
```diff
@@ -25,8 +26,16 @@ class Bernoulli(Distribution):
     """
     has_enumerate_support = True
 
-    def __init__(self, probs):
-        self.probs, = broadcast_all(probs)
+    def __init__(self, probs=None, logits=None):
+        if (probs is None) == (logits is None):
+            raise ValueError("Got probs={}, logits={}. Either `probs` or `logits` must be specified, "
+                             "but not both.".format(probs, logits))
```
Comment (apaszke) [inline torch/distributions/bernoulli.py]: I'm afraid that this will print a huge message containing two tensors. I think it's enough to just say that exactly one of them should be specified.

File: test/test_distributions.py [hunk 3]
```diff
@@ -225,8 +226,14 @@ def ref_log_prob(idx, val, log_prob):
             self.assertEqual(log_prob, math.log(prob if val else 1 - prob))
 
         self._check_log_prob(Bernoulli(p), ref_log_prob)
+        self._check_log_prob(Bernoulli(logits=p.log() - (-p).log1p()), ref_log_prob)
```
Comment (apaszke) [inline test/test_distributions.py]: why not just `p.log()`?

File: test/test_distributions.py [hunk 4]
```diff
@@ -1001,5 +1013,108 @@ def test_laplace_shape_tensor_params(self):
...
+                        expected_gradient=None,
+                        prec=1e-5):
+        if probs is not None:
+            p = Variable(probs, requires_grad=True)
+            dist = dist_class(p)
+        else:
+            p = Variable(logits, requires_grad=True)
+            dist = dist_class(logits=p)
+        log_pdf = dist.log_prob(Variable(x))
+        log_pdf.retain_grad()
```
Comment (apaszke) [inline test/test_distributions.py]: You don't seem to be using `log_pdf.grad`, and even if you were it will always be all ones

File: torch/nn/functional.py [hunk 5]
```diff
@@ -1273,7 +1273,7 @@ def binary_cross_entropy(input, target, weight=None, size_average=True, reduce=T
     return torch._C._nn.binary_cross_entropy(input, target, weight, size_average, reduce)
 
 
-def binary_cross_entropy_with_logits(input, target, weight=None, size_average=True):
+def binary_cross_entropy_with_logits(input, target, weight=None, size_average=True, reduce=True):
```
Comment (apaszke) [inline torch/nn/functional.py]: Can you add `reduce` to the docstring?

**Review by apaszke — COMMENTED** [review]
That looks good. My suggestion would be to make `probs` and `logits` properties that are lazily instantiated - there's no point in putting in work to do all these expensive operations (log, exp, ...) in the constructor if the only thing someone needs is to draw some samples given probabilities. If you're concerned about making these attribute lookups slower, we can always use [non-data descriptors](https://docs.python.org/2/howto/descriptor.html#descriptor-protocol) to create them. I can explain that in detail if you're interested.

File: test/test_distributions.py (hunk 3 above)
Comment (fritzo) [reply test/test_distributions.py]: There are two common definitions of logits, one corresponding to log-likelihood and one corresponding to log-odds. The `logits` arg to `Categorical` is the former. The `logits` arg to `Bernoulli` is the latter and is equal to `log(p / (1-p))`. If we were to implement `Bernoulli` via `Categorical`, the logits would be a pair of numbers `[log(p), log(1-p)]` but `Bernoulli` can simply use the log-ratio of the two.

File: torch/nn/functional.py (hunk 5 above)
Comment (neerajprad) [reply torch/nn/functional.py]: done.

File: test/test_distributions.py (hunk 4 above)
Comment (neerajprad) [reply test/test_distributions.py]: My bad, this was left over from some testing I was doing earlier. Removed.

File: torch/distributions/bernoulli.py (hunk 2 above)
Comment (neerajprad) [reply torch/distributions/bernoulli.py]: Done.

**Comment by neerajprad** [comment]
>  If you're concerned about making these attribute lookups slower, we can always use non-data descriptors to create them.

Good idea! I will look into this and update the PR.

File: test/test_distributions.py (hunk 3 above)
Comment (apaszke) [reply test/test_distributions.py]: That makes sense, but I think it would be good to use a different name (perhaps `log_odds`), so that it's immediately clear that it's expected to be a different thing. I can merge it as is if you think that's unnecessary, but it feels like a safer interface to me.

Comment (neerajprad) [reply test/test_distributions.py]: I was trying to be consistent with how Tensorflow defines it, but we could change it to log odds for Bernoulli too.

**Comment by neerajprad** [comment]
> That looks good. My suggestion would be to make probs and logits properties that are lazily instantiated 

Created a lazy property decorator in `distributions/utils.py` for this.

File: test/test_distributions.py (hunk 3 above)
Comment (apaszke) [reply test/test_distributions.py]: I don't think keeping it 100% consistent is worth it, especially that it's not a silent difference. It will be immediately clear that you have to use a different name, or you'll get an error.

I'm a bit worried of calling two semantically different things using the same name (if you don't at least have a type system), because that's a recipe for an easy failure that will take a long time to debug.

File: torch/distributions/bernoulli.py [hunk 6]
```diff
@@ -31,16 +31,22 @@ def __init__(self, probs=None, logits=None):
             raise ValueError("Either `probs` or `logits` must be specified, but not both.")
         if probs is not None:
             self.probs, = broadcast_all(probs)
-            self.logits = probs_to_logits(self.probs, is_binary=True)
         else:
             self.logits, = broadcast_all(logits)
-            self.probs = logits_to_probs(self.logits, is_binary=True)
-        if isinstance(probs, Number):
+        if isinstance(probs, Number) or isinstance(logits, Number):
```
Comment (apaszke) [inline torch/distributions/bernoulli.py]: One suggestion (feel free to ignore that if you don't want to implement it), for both here and in `Categorical` would be to define `probs_or_logits` in the `if` above, to make it point to whatever was specified to the distribution. You're only using the shape so that's enough, it will save you all the conditionals, and we will be sure that you're not triggering the lazy attributes by accident.

**Review by apaszke — COMMENTED** [review]
Nice work with lazy attrs!

File: test/test_distributions.py (hunk 3 above)
Comment (neerajprad) [reply test/test_distributions.py]: Actually, `logits` and `log odds` refer to the same thing, namely, `log (p/(1-p))`. The issue is that for the multidimensional case, `logits` is used to loosely refer to log of probability (possibly unnormalized). So rather than changing `logits` to `log_odds` in Bernoulli, we could change the `logits` in Categorical to `log_probs` instead (that sounds similar to the `log_prob` method though) to resolve this potential confusion.

@fritzo - do you have any thoughts / preferences on the naming?

File: torch/distributions/bernoulli.py (hunk 6 above)
Comment (neerajprad) [reply torch/distributions/bernoulli.py]: Good idea..let me refactor this part.

File: test/test_distributions.py (hunk 3 above)
Comment (fritzo) [reply test/test_distributions.py]: Tensorflow uses `logits` informally throughout to refer to both log-odds and log-probs. In an effort to maximize compatibility with [tf.distributions.Categorical](https://www.tensorflow.org/api_docs/python/tf/distributions/Categorical), I recommend we name this parameter `logits`.

Comment (apaszke) [reply test/test_distributions.py]: Alright, let's leave it as is

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Ready to merge, but can you please resolve the conflicts?

**Comment by neerajprad** [comment]
@apaszke - We need to whitelist this for testing first, as we have only run lint check on this PR.

File: test/test_distributions.py [hunk 7]
```diff
@@ -1055,5 +1068,107 @@ def test_laplace_shape_tensor_params(self):
...
+
+            self._test_pdf_score(dist_class=Bernoulli,
+                                 probs=tensor_type([1e-4]),
+                                 x=tensor_type([1]),
+                                 expected_value=tensor_type([math.log(1e-4)]),
+                                 expected_gradient=tensor_type([10000]))
+
+            # Lower precision due to:
+            # >>> 1 / (1 - torch.FloatTensor([0.9999]))
+            # 9998.3408
```
Comment (neerajprad) [inline test/test_distributions.py]: @apaszke - I was a bit surprised by this. Any ideas why this does not return `10,000`, as you might expect?

Comment (apaszke) [reply test/test_distributions.py]: Because 1 - 0.9999 = 9.999999999998899e-05 (that's in double precision, but pretty much the same thing happens in fp32).

Comment (neerajprad) [reply test/test_distributions.py]: Thanks. What I found interesting for `FloatTensor` (`DoubleTensor` returns expected result to first five decimal places at least) was that the result is less than `10,000` for `0.9999` and beyond. I am probably missing some floating point arithmetic details here! 

e.g.
```
>>> 1/(1-torch.FloatTensor([0.99]))
 100.0001
[torch.FloatTensor of size 1]

>>> 1/(1-torch.FloatTensor([0.999]))
 1000.0129
[torch.FloatTensor of size 1]

>>> 1/(1-torch.FloatTensor([0.9999]))
 9998.3408
[torch.FloatTensor of size 1]

```

**Comment by neerajprad** [comment]
Oops..more conflicts, will rebase.

**Comment by neerajprad** [comment]
@apaszke - good to merge now.
